packages/playwright-core/src/server/injected/recorder.ts

```diff
@@ -82,13 +80,20 @@ export class Recorder {
     this._highlight.install();
   }
 
-  setUIState(state: UIState) {
+  setUIState(state: UIState, delegate: RecorderDelegate) {
+    this._delegate = delegate;
+
+    if (state.mode !== 'none' || state.actionSelector)
+      this.refreshListenersIfNeeded();
+    else
+      removeEventListeners(this._listeners);
```

Comment:
This must uninstall the highlight, otherwise line 58 will prevent the listeners from being added next time we change `state.mode`.